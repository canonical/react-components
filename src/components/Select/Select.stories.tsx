import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Select from "./Select";

const meta: Meta<typeof Select> = {
  component: Select,
  tags: ["autodocs"],

  argTypes: {
    caution: {
      control: {
        type: "text",
      },
    },

    error: {
      control: {
        type: "text",
      },
    },

    help: {
      control: {
        type: "text",
      },
    },

    label: {
      control: {
        type: "text",
      },
    },

    success: {
      control: {
        type: "text",
      },
    },

    id: {
      control: {
        disable: true,
      },
    },

    options: {
      control: {
        disable: true,
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Select>;

export const Default: Story = {
  name: "Select",

  args: {
    name: "exampleSelect",
    id: "exampleSelect2",
    defaultValue: "",

    options: [
      {
        value: "",
        disabled: true,
        label: "Select an option",
      },
      {
        value: "1",
        label: "Cosmic Cuttlefish",
      },
      {
        value: "2",
        label: "Bionic Beaver",
      },
      {
        value: "3",
        label: "Xenial Xerus",
      },
    ],

    label: "Ubuntu releases",
  },
};

export const SelectMultiple: Story = {
  render: () => (
    <Select
      name="exampleSelectMulti"
      id="exampleSelectMulti2"
      options={[
        {
          value: "",
          disabled: true,
          label: "Select...",
        },
        {
          value: "1",
          label: "Cosmic Cuttlefish",
        },
        {
          value: "2",
          label: "Bionic Beaver",
        },
        {
          value: "3",
          label: "Xenial Xerus",
        },
      ]}
      label="Ubuntu releases"
      multiple
    />
  ),

  name: "Select multiple",
};
