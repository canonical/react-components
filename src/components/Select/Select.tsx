import classNames from "classnames";
import React, { useEffect, useId, useRef } from "react";
import type {
  ChangeEventHandler,
  ReactNode,
  OptionHTMLAttributes,
  SelectHTMLAttributes,
} from "react";

import Field from "../Field";

import type { ClassName, PropsWithSpread } from "types";

type Option = OptionHTMLAttributes<HTMLOptionElement>;

/**
 * The props for the Select component.
 */
export type Props = PropsWithSpread<
  {
    /**
     * The content for caution validation.
     */
    caution?: ReactNode;
    /**
     * Optional class(es) to pass to the input element.
     */
    className?: ClassName;
    /**
     * The content for error validation.
     */
    error?: ReactNode;
    /**
     * Help text to show below the field.
     */
    help?: ReactNode;
    /**
     * The id of the input.
     */
    id?: string | null;
    /**
     * The label for the field.
     */
    label?: ReactNode;
    /**
     * Optional class(es) to pass to the label component.
     */
    labelClassName?: string | null;
    /**
     * Function to run when select value changes.
     */
    onChange?: ChangeEventHandler<HTMLSelectElement> | null;
    /**
     * Array of options that the select can choose from.
     */
    options?: Option[] | null;
    /**
     * Whether the field is required.
     */
    required?: boolean;
    /**
     * Whether the form field should have a stacked appearance.
     */
    stacked?: boolean;
    /**
     * The content for success validation.
     */
    success?: ReactNode;
    /**
     * Whether to focus on the input on initial render.
     */
    takeFocus?: boolean;
    /**
     * Optional class(es) to pass to the wrapping Field component
     */
    wrapperClassName?: ClassName;
  },
  SelectHTMLAttributes<HTMLSelectElement>
>;

const generateOptions = (options: Props["options"]) =>
  options?.map(({ label, value, ...props }) => (
    <option value={value} key={`${value}` || label} {...props}>
      {label}
    </option>
  ));

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Select](https://docs.vanillaframework.io/base/forms/#select).
 *
 * Use the Select component to create a drop-down list.
 */
const Select = ({
  caution,
  className,
  error,
  help,
  id,
  label,
  labelClassName,
  onChange,
  options,
  required,
  stacked,
  success,
  takeFocus,
  wrapperClassName,
  ...selectProps
}: Props): JSX.Element => {
  const selectRef = useRef(null);
  const validationId = useId();
  const defaultSelectId = useId();
  const selectId = id || defaultSelectId;
  const helpId = useId();
  const hasError = !!error;

  useEffect(() => {
    if (takeFocus) {
      selectRef.current.focus();
    }
  }, [takeFocus]);

  return (
    <Field
      caution={caution}
      className={wrapperClassName}
      error={error}
      forId={selectId}
      help={help}
      helpId={helpId}
      isSelect={true}
      label={label}
      labelClassName={labelClassName}
      required={required}
      stacked={stacked}
      success={success}
      validationId={validationId}
    >
      <select
        aria-describedby={[help ? helpId : null, success ? validationId : null]
          .filter(Boolean)
          .join(" ")}
        aria-errormessage={hasError ? validationId : null}
        aria-invalid={hasError}
        className={classNames("p-form-validation__input", className)}
        id={selectId}
        onChange={(evt) => onChange && onChange(evt)}
        ref={selectRef}
        {...selectProps}
      >
        {generateOptions(options)}
      </select>
    </Field>
  );
};

export default Select;
