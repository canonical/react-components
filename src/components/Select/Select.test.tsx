import { mount, shallow } from "enzyme";
import React from "react";

import Select from "./Select";

describe("Select ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <Select
        id="test-id"
        options={[
          { value: "", disabled: true, label: "Select an option" },
          { value: "1", label: "Cosmic Cuttlefish" },
          { value: "2", label: "Bionic Beaver" },
          { value: "3", label: "Xenial Xerus" },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onChange prop", () => {
    const onChangeMock = jest.fn();
    const event = {
      event: "test-event",
    };
    const component = shallow(<Select options={[]} onChange={onChangeMock} />);
    component.find("select").simulate("change", event);
    expect(onChangeMock).toBeCalledWith(event);
  });

  it("can add additional classes", () => {
    const wrapper = shallow(<Select className="extra-class" options={[]} />);
    const className = wrapper.find("select").prop("className");
    expect(className.includes("p-form-validation__input")).toBe(true);
    expect(className.includes("extra-class")).toBe(true);
  });

  it("can take focus on first render", () => {
    const container = document.createElement("div");
    document.body.appendChild(container);
    const wrapper = mount(<Select options={[]} takeFocus />, {
      attachTo: container,
    });
    expect(wrapper.find("select").getDOMNode()).toBe(document.activeElement);
  });

  it("can take null options", () => {
    const wrapper = shallow(<Select options={null} />);
    expect(wrapper.exists()).toBe(true);
  });
});
