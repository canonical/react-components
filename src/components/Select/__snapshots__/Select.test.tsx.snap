// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders 1`] = `
<Field
  forId="test-id"
  helpId="mock-nanoid-2"
  isSelect={true}
  validationId="mock-nanoid-1"
>
  <select
    aria-describedby={null}
    aria-errormessage={null}
    aria-invalid={false}
    className="p-form-validation__input"
    id="test-id"
    onChange={[Function]}
  >
    <option
      disabled={true}
      key="Select an option"
      value=""
    >
      Select an option
    </option>
    <option
      key="1"
      value="1"
    >
      Cosmic Cuttlefish
    </option>
    <option
      key="2"
      value="2"
    >
      Bionic Beaver
    </option>
    <option
      key="3"
      value="3"
    >
      Xenial Xerus
    </option>
  </select>
</Field>
`;
